Adicionar populador do DB + refatoração de scripts

th-duvanel opened the pull request:
Ei, pessoal, depois de um tempo, finalmente consegui fazer o script populador do banco de dados.
Vamos conferir o que fiz na última semana, além das modificações na configuração atual do repositório:

# Estrutura

1. Removi a pasta `/scripts/data`, já que não teremos mais nenhum script além dos relacionados a dados.
2. Adicionei a pasta `/cache/` (apenas na nossa imaginação, porque o Git não permite o commit de pastas vazias, então não precisamos colocar o nome dela no gitignore).
3. Adicionei a pasta `/src/db/` para separar as configurações do banco de dados.

# Scripts

1. Adicionei o script populador do banco de dados (obviamente).
2. Removi as tabelas de classes do script processador de CSV do @OJarrisonn e movi para a pasta `/src/db/`.
3. Fiz algumas modificações no meu script `fetcher.py` para atender às novas configurações.
4. Fiz algumas modificações no script `process_sus_data.py` para atender ao novo populador.
5. Finalmente, adicionei o requisito do SQLAlchemy.

Ainda não testei o suficiente porque estou pensando em modificar algum código e gostaria de saber a opinião de vocês sobre isso. Vou fazer outro PR para modificar o `Dockerfile` e incluir o novo script populador, já que precisamos usá-lo dentro do container.

th-duvanel committed:
refactor: added /cache/ folder

th-duvanel committed:
refactor: removed data folder

th-duvanel committed:
feat: added database populator script (without general_info)

th-duvanel committed:
feat: added sqlachemy requirement

th-duvanel committed:
refactor: added sql tables for db classes

th-duvanel committed:
refactor: removed tables from csv processor

th-duvanel committed:
refactor: changed download_stablishments to download_stablishment (only one at time)

th-duvanel committed:
feat: added GeneralInfo populator

nahharris commented on scripts/populator.py:
```diff
@@ -0,0 +1,112 @@
+DB_NAME = os.getenv("DBNAME")
+DB_PASSWORD = os.getenv("DBPASSWORD")
+
+DB_URL = f"postgresql://{DB_USER}:{DB_PASSWORD}@localhost:5432/{DB_NAME}"
```
Futuramente, pode ser necessário adicionar uma variável para não usar esse `localhost` hardcoded. Dentro da composição do docker esse script só funcionaria se fosse executado dentro da docker que possui o banco

th-duvanel replied:
Concordo. Coloquei esse URL apenas para ter alguma coisa para acessá-lo, mas com certeza era bom mudar. Aceito sugestões.

No caso, era uma boa por como ENVvar tbm?

nahharris replied:
Isso

nahharris commented on scripts/populator.py:
```diff
@@ -0,0 +1,112 @@
+
+    for cnes_code in elasticnes['CNES']:
+        download_stablishment(cnes_code)
+        GeneralInfo = general_info(elasticnes, f"{DATA_PATH}/{cnes_code}.json")
+        populate_db_from_object(GeneralInfo)
```
`GeneralInfo` já é o nome de uma classe. Como isso é um valor, opte por `general_info`

`PascalCase` para tipos, `snake_case` para variáveis e funções

th-duvanel replied:
`general_info` também é o nome da função do script de processar os dados. Irei alterar o nome das funções para process_*.

nahharris commented on scripts/populator.py:
```diff
@@ -0,0 +1,112 @@
+    Returns:
+    None
+    """
+    ServiceRecords = service_records(f"{DATA_PATH}{ESPEC_FILE_NAME}")
+    for record in ServiceRecords:
```
Opte por `records` ao invés de `ServiceRecords`

th-duvanel replied:
Ok, já alterado.

th-duvanel committed:
refactor: minor variable names fix

th-duvanel committed:
refactor: change functions name

nahharris commented:
Por uma questão de implementação eu gostaria de solicitar uma nova tabela. Essa tabela seria uma lista de strings únicas contendo todas as especialidades. Para povoar essa tabela basta pegar a propriedade `service` de todos `ServiceRecord`, removendo as repetições.

Com isso a rota `/especialidades` será muito mais eficiente

th-duvanel commented:
> Por uma questão de implementação eu gostaria de solicitar uma nova tabela. Essa tabela seria uma lista de strings únicas contendo todas as especialidades. Para povoar essa tabela basta pegar a propriedade `service` de todos `ServiceRecord`, removendo as repetições.
> 
> Com isso a rota `/especialidades` será muito mais eficiente

Boa ideia. Posso fazer isso rapidinho.
Inclusive, faltou algo bem importante: filtrar apenas as unidades de SP. Vamos poder remover a coluna de município, e é algo bem fácil de fazer.

th-duvanel committed:
feat: adds MedicalService class and table

th-duvanel committed:
feat: adds filter to SP city + adds medical_service populator

th-duvanel committed:
feat: adds medical services processor

nahharris commented on scripts/populator.py:
```diff
@@ -0,0 +1,150 @@
+from sqlalchemy.ext.declarative import declarative_base
+from sqlalchemy.orm import sessionmaker
+
+from src.db.tables import Base, GeneralInfo, ServiceRecord, MedicalService
```
Esse import está funcionando?

th-duvanel replied:
Para funcionar o import é necessário modificar a variável de ambiente `PYTHONPATH` para o diretório do projeto ser tratado como raiz. Ela terá que ser modificada logo após o Docker iniciar. Vou modificar isso no Dockerfile.

th-duvanel replied:
Para funcionar, é necessário a PYTHON_PATH estar definida. No docker o ideal é ela já estar definida.

nahharris commented on scripts/populator.py:
```diff
@@ -0,0 +1,150 @@
+
+    populate_service_records()
+    populate_medical_services()
+    populate_general_info()
```
Poderia ser mais interessante que essa variável elasticnes não fosse global, mas sim passada como argumento para as funções que necessitam

nahharris commented on scripts/populator.py:
```diff
@@ -0,0 +1,150 @@
+    """
+    medical_services = process_medical_services(elasticnes)
+
+    for service in medical_service:
```
Deveria ser `medical_services`

th-duvanel committed:
feat: adds db address and the app as python path

th-duvanel committed:
refactor: elasticnes dataframe as a argument, not as a global var

th-duvanel committed:
feat: adds psycopg2 as a requirement

th-duvanel committed:
feat: adds the city and state

th-duvanel committed:
fix: table types were incorrect

th-duvanel commented:
Consertei vários problemas e segui as recomendações do @OJarrisonn.

Como tem vários commits, o ideal era fazer um squash deles ao mergear, senão via ficar muito repetitivo, já que eu coloquei um commit por arquivo modificado.

th-duvanel committed:
fix: removes scripts from root

th-duvanel committed:
feat: adds psycopg2 binary requirement

th-duvanel committed:
fix: changes the envvar

th-duvanel commented:
Olá a todos,
Temos um problema: o .csv que imaginávamos que ia ser suficiente, não é. Ao baixar os arquivos de dados "genéricos" como meu script faz, ele baixa um zip com várias informações separadas.

Temos duas opções para seguir: 

1. Baixar na marra o arquivo .csv já filtrado,  o que não necessitaria de modificação no script mas removeria totalmente a "automateza" dele
2. Modificar totalmente as tabelas, uma para cada arquivo, e fazer as rotas se virarem.

Independente da escolha, eu não vou continuar fazendo essas tasks sozinho por terem uma complexidade mais escalável. O certo é seguir a segunda opção, bastaria completar meu script populator com mais tabelas e suas respectivas funções populadoras, além, claro, de selecionar os arquivos necessários.

Gostaria de saber a opinião de vocês sobre como vamos seguir e como vamos separar essas tarefas para cada um. Eu até poderia extender essa issue para outra sprint e fazer sozinho, mas imagino que vocês estão dependendo de mim para poder fazer todas as rotas.

Fico no aguardo de vocês para seguir o próximo passo.
@OJarrisonn @jppaulo06

th-duvanel committed:
fix: environemnt variable fixes

th-duvanel committed:
feat: adds python path for the back-end environment

th-duvanel committed:
feat: adds curl dependency for back-end container

th-duvanel committed:
fix: correct cache path applied

nahharris commented:
> Independente da escolha, eu não vou continuar fazendo essas tasks sozinho por terem uma complexidade mais escalável. O certo é seguir a segunda opção, bastaria completar meu script populator com mais tabelas e suas respectivas funções populadoras, além, claro, de selecionar os arquivos necessários.

Sem problema, eu assumo essa no próximo ciclo

th-duvanel committed:
wip: script fixing for the new way to download data

nahharris committed:
fix: now properly populates

TODO: some json files might fail to be parsed somehow

nahharris committed:
fix: the downloader will persist

Sometimes downloading of `.json` from adasus may fail. The script will
try again until it succeeds

Signed-off-by: OJarrisonn <j.h.m.t.v.10@gmail.com>

nahharris commented:
O script já está pronto para uso, mas com algumas ressalvas:

- O `.csv` que vem do elasticnes deve ser baixado manualmente e salvo como `data/DADOS_CNES.csv`
- Alguns downloads do adasus falham em duas formas diferentes: o servidor rejeita a conexão ou o servidor envia uma paǵina de erro. No primeiro caso tentamos o download novamente, no segundo caso ignoramos a unidade
- É necessário executar `drop table` manualmente no banco de dados antes de executar o script

Mudanças sugeridas para uma próxima versão:

- Baixar automaticamente o csv do elasticnes
- Usar um modelo transacional reversível para apagar as tabelas, popular novamente e só então _commitar_  as mudanças concluindo a transação
- Verificar apropriadamente se o download do adasus foi bem sucedido e tentar novamente caso falhe (no segundo caso)
- Apagar código não utilizado

@jppaulo06 gostaria que você tentasse executar esse script e aprovasse esse PR

```
# verifique se você tem o arquivo na pasta `data/` e que seu banco de dados está vazio
docker compose --profile dev up --build
# entre no container do servidor com docker exec
python3 scripts/populator.py
```

nahharris commented on scripts/populator.py:
```diff
@@ -0,0 +1,175 @@
+    elasticnes = pd.read_csv(f"{DATA_PATH}DADOS_CNES.csv")
+
+    print("[2] Populating medical services")
+    #populate_medical_services(elasticnes)
```
Tem de descomentar essa linha @OJarrisonn

nahharris commented on scripts/populator.py:
```diff
@@ -0,0 +1,175 @@
+    print("[2] Populating medical services")
+    #populate_medical_services(elasticnes)
+    print("[3] Populating service records")
+    #populate_service_records(elasticnes)
```
Essa também @OJarrisonn

jppaulo06 committed:
WIP
